### ISSUE TITLE
JSON content is mixed up between nodes

### ISSUE THREAD

**Issue by runemadsen** [gatsbyjs/gatsby#33432]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

I've been struggling for weeks to figure out what's going on with this weird bug. I have a series of `.json` files that I'm loading with `gatsby-source-filesystem` and `gatsby-transformer-json` (The files have english or spanish text because I'm setting up a translation system, but that doesn't really matter for this bug). It's a very simple setup that I've done many times before.

However, when I load the JSON nodes through GraphQL, gatsby has replaced some of the node content with content from other nodes. I don't know how else to describe it. As an example, take these two JSON files:

```
// content/$1_Unistroke_Recognizer.json
{
  "sentence" : "Implementation of the ....."
  ...
}

// content/Unistroke_Spanish.json
{
  "sentence" : "Implementación del  ....."
  ...
}
```

And the GraphQL query:

```
allFile(
    filter: {sourceInstanceName: {eq: "contributions"}}
  ) {
    nodes {
      name
      childJson {
        sentence
      }
    }
  }
```

And the result:

```
{
  "data": {
    "allFile": {
      "nodes": [
        ...
        {
          "name": "$1_Unistroke_Recognizer",
          "childJson": {
            "sentence": "Implementación del ..."
          }
        },
        {
          "name": "Unistroke_Spanish",
          "childJson": {
            "sentence": "Implementación del ..."
          }
        }
      ]
    }
  },
  "extensions": {}
}
```

You can see that the Spanish version is in both nodes. I have put debug statements in my `gatsby-node` file, and the JSON nodes are created properly. I have also tried every possible thing to figure out what's going on, but I just can't get my head around why this is happening. Furthermore, when I clean the cache and rerun the dev server, sometimes it's the english sentence in both, sometimes it's correct.

I left a few other files in the test repo, so you can get a full picture of how this is happening across everything.

Can anyone please give me any pointers to what's going on? Note that the reproduction URL is for a branch with a very slim repo, so it's easy to test. The `main` branch is for a much larger Gatsby site running the processing.org website.

### Reproduction Link

https://github.com/processing/processing-website/tree/debug-json-translation-merge

### Steps to Reproduce

1. Clone down the repo
2. Run `git checkout debug-json-translation-merge`
3. Run `npm i`
4. Run `npm run dev`
5. Open `localhost:8000/__graphql` and run the following query:

```
{
  allFile(
    filter: {sourceInstanceName: {eq: "contributions"}}
  ) {
    nodes {
      name
      childJson {
        sentence
      }
    }
  }
}
```

### Expected Result

That the resulting JSON nodes have the content that corresponds to the content in the actual files

### Actual Result

The JSON nodes do not have the content that is in their corresponding files

### Environment

```shell
System:
    OS: macOS 10.15.7
    CPU: (8) x64 Intel(R) Core(TM) i7-1068NG7 CPU @ 2.30GHz
    Shell: 5.7.1 - /bin/zsh
  Binaries:
    Node: 14.17.6 - ~/.nvm/versions/node/v14.17.6/bin/node
    npm: 6.14.15 - ~/.nvm/versions/node/v14.17.6/bin/npm
  Languages:
    Python: 2.7.16 - /usr/bin/python
  Browsers:
    Chrome: 94.0.4606.61
    Firefox: 91.0.2
    Safari: 14.0
  npmPackages:
    gatsby: ^3.14.0 => 3.14.0 
    gatsby-source-filesystem: ^3.14.0 => 3.14.0 
    gatsby-transformer-json: ^3.14.0 => 3.14.0
```

### Config Flags

DEV_SSR: true,
FAST_DEV: false,
PRESERVE_FILE_DOWNLOAD_CACHE: false

**Comment by runemadsen** [comment]
So @Anadroid helped me debug this, and we realized that the error was that the JSON data came from another repo and they had an `id` property in them. Apparently, this `id` property gets used as the node ID, and when several JSON files have the same `id`, Gatsby will basically use whichever gets loaded last in the async build process.

So in this case, we had copied a JSON file to translate it, preserving the same `id`, and therefore the content was mixed between the two files. 

Is this the way it's supposed to work?
